rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,342 @@
+the batch in multiple parts to reduce the amount of data buffered
+locally.
+
+1. The sink operates as before, where objects are created based only
+   on the existing batch settings. This will initially be the default
+   configuration to avoid breaking existing setups but will be
+   deprecated.
+1. The sink buffers batches until it can generate a complete part, and
+   assembles the parts into an object only at the end of an
+   interval. None of the intermediate parts are visible until the
+   final object is assembled.
+
+This change will introduce the following new configuration items for
+this sink, which will all default to operating the same as the
+existing sink.
+
+- `data_dir` (string): The path to the directory in which the state
+  data is stored. If this is not set, it is derived from the global
+  `data_dir` setting.
+- `batch.min_bytes` (integer): This specifies the target size a
+  locally-stored part must reach before it is uploaded. Due to the
+  limitations of the S3 API, this has an minimum value of 5MB, which
+  is also the default value when `upload_interval` is set.
+- `upload_interval` (string): When should the sink
+  complete the upload process and initiate a new one. If this is
+  unset, the existing batch-based object creation mechanism is
+  used. the following values are permitted:
+  - `"hourly"`: Complete the upload at the end of the hour exactly.
+  - `"daily"`: Complete the upload at midnight local time each day.
+  - `"at TIME"`: Complete the upload at a specific time each day.
+
+### Implementation
+
+#### Convert the S3 sink to use a new partitioned `Batcher`
+
+The `Batcher` interface, used by a select few new-style stream sinks,
+is a more generic batch collection setup. It provides for the
+semantics required for this implementation, where events must be
+encoded before batching in order to calculate their byte
+size. However, it does not support partitioning, which is only
+provided by the `PartitionBatcher` which does not support encoded size
+limiting. We will need to create a new partitioned batcher combining
+these two capabilities.
+
+```rust
+pub struct PartitionBatcher<S, C, P> {
+    state: C,
+    stream: Fuse<S>,
+    partitioner: P,
+    timer: Maybe<Sleep>,
+}
+
+impl<S, C, P> Stream for PartitionBatcher<S, C, P>
+where
+    S: Stream<Item = P::Item>,
+    C: BatchConfig<S::Item>,
+    P: Partitioner,
+{
+    type Item = (Prt::Key, C::Batch);
+}
+```
+
+The S3 sink will then be converted to use the new partitioned batcher,
+with the additional behavior change of encoding events before batching
+rather than after.
+
+#### Add a minimum batch size limit to the batcher
+
+Since the multipart object creation API requires each part to have a
+minimum size, this batcher will also have to enforce a minimum size
+before an upload is initiated. The `Batcher` interface will be
+modified to respect this interface and not emit batches until one is
+ready (except at shutdown).
```

Comment:
Would we be looking to "complete" the object when shutting down? My understanding from what you stated above only the "last" batch could be under the min size.